sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,329 @@
+      break;
+    case kDifHmacModeSha256:
+      mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                        HMAC_CFG_HMAC_EN);
+      break;
+    default:
+      return kDifHmacError;
+  }
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_message_endianness_set(
+    const dif_hmac_t *hmac, dif_hmac_endianness_t endianness) {
+  if (hmac == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  switch (endianness) {
+    case kDifHmacEndiannessBig:
+      mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                      HMAC_CFG_ENDIAN_SWAP);
+      break;
+    case kDifHmacEndiannessLittle:
+      mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                        HMAC_CFG_ENDIAN_SWAP);
+      break;
+    default:
+      return kDifHmacError;
+  }
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_digest_endianness_set(
+    const dif_hmac_t *hmac, dif_hmac_endianness_t endianness) {
+  if (hmac == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  switch (endianness) {
+    case kDifHmacEndiannessBig:
+      mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                      HMAC_CFG_DIGEST_SWAP);
+      break;
+    case kDifHmacEndiannessLittle:
+      mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                        HMAC_CFG_DIGEST_SWAP);
+      break;
+    default:
+      return kDifHmacError;
+  }
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_key_set(const dif_hmac_t *hmac, const uint8_t *key) {
+  if (hmac == NULL || key == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  load_key(hmac, key);
+
+  return kDifHmacOk;
+}
+
+dif_hmac_fifo_result_t dif_hmac_fifo_push(const dif_hmac_t *hmac,
+                                          const void *data, size_t len,
+                                          size_t *bytes_sent) {
+  if (hmac == NULL || data == NULL) {
+    return kDifHmacFifoBadArg;
+  }
+
+  const uint8_t *data_sent = (const uint8_t *)data;
+  int bytes_remaining = len;
+
+  while (bytes_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    bool word_aligned = (uintptr_t)data_sent % sizeof(uint32_t) == 0;
+    int bytes_written = 0;
+
+    if (bytes_remaining < sizeof(uint32_t) || !word_aligned) {
+      // Individual byte writes are needed if the buffer isn't aligned or there
+      // are no more full words to write.
+      mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
+      bytes_written = 1;
+    } else {
+      // |data_sent| is word-aligned and there are still words to write.
+      uint32_t word = read_32(data_sent);
+      mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
+      bytes_written = sizeof(uint32_t);
+    }
+
+    bytes_remaining -= bytes_written;
+    data_sent += bytes_written;
+  }
```

Comment:
This also looks like it wants `mmio_region_memcpy()`.

Reply:
I think doing so will be incorrect here.

`region_memcpy` will do a read modify write on the last write, ensuring the last write is 32 bits wide, regardless of the width of the original buffer (so the length of the total written buffer becomes a multiple of 4 bytes). However, in HMAC, the packer logic is calculating the length based on the writes it is seeing, so doing a single byte-write is actually a different operation to doing a word-write (one increases the hashed length by 1, the other by 4.

Reply:
[I think so too](https://github.com/lowRISC/opentitan/pull/2066#issuecomment-622673045).